tests/controller_tests/test_ingest.py

```diff
@@ -515,12 +515,33 @@ def test_extract_data_extracts_from_multiple_sheets(example_data_model_file):
     assert isinstance(list(workbook.values())[0], pd.DataFrame)
 
 
-def test_next_submission_id_first_submission(test_client):
-    sub_id = next_submission_id(reporting_round=1)
-    assert sub_id == "S-R01-1"
-
-
-def test_next_submission_id_existing_submissions(test_client):
+@pytest.mark.parametrize(
+    "reporting_round, expected_submission_id",
+    [
+        (1, "S-R01-ABCDEF"),
+        (2, "S-R02-ABCDEF"),
+        (3, "S-R03-ABCDEF"),
+        (4, "S-R04-ABCDEF"),
+    ],
+)
+def test_next_submission_id_first_submission(reporting_round, expected_submission_id, monkeypatch):
+    # Monkeypatch the random key generator to return a fixed value
+    monkeypatch.setattr("random.choices", lambda population, k: ["A", "B", "C", "D", "E", "F"])
+
+    sub_id = next_submission_id(reporting_round)
+    assert sub_id == expected_submission_id
```

Comment:
I don't think we even need to do this patching, could we just assert that the start of the string is as expected, and that the end is 6 caps letteers?

Reply:
i.e. if we parameterise the func with just the rounds?

Reply:
```
assert sub_id.startswith(f"S-R0{1}".format(reporting_round)
assert sub_id[:-6].isupper()
assert len(sub_id) == 11
```
Something like that? To me it seems a little messier. Is there a reason we want to avoid patching random.choice?

Reply:
There's no reason to avoid it par se, but this kind of pattern, ie mock something, then assert it is what we just mocked it to be, doesn't really achieve much, especially WRT testing what the function actually does. The most important things to test here are:

- The bit at the start is generated as per the logic, based on the round parameter passed
- The string format looks correct
- The final chars are generated as expected ie 6 capital chars.

The last 2 are not really tested if we just tell the function what to return in a mock. And the function doesn't really produce a definitive id (as they're random) so why test that it does?

Reply:
> ```
> assert sub_id.startswith(f"S-R0{1}".format(reporting_round)
> assert sub_id[:-6].isupper()
> assert len(sub_id) == 11
> ```
> 
> Something like that? To me it seems a little messier. Is there a reason we want to avoid patching random.choice?

Yeah basically. It's not really "messier" - it's asserting the structure of the string is generated as we expect. If we just mock the expected outcome, then assert against that, if something changes in the code that should break the test, the test may still pass as it effectively ignores the pertinent part of the code that's supposed to be tested

Reply:
also, I think the code would be
```
assert sub_id.startswith(f"S-R0{reporting_round}")
assert sub_id[-6:].isupper()
assert len(sub_id) == 11
```

Reply:
But you could always split into suitable var names if you want it to be clearer what it's testing?

Reply:
Should be fine to just make those three assertions, it's a relatively straightforward method to test